### PR TITLE
Cross platform paths

### PR DESCRIPTION
**Pull request by jranke**
This PR  fixes a test failure on windows. The commit that fixes the test is actually only the second one.

The first commit removes some complexity in `compileCode`. It removes handling of some corner case where `R CMD SHLIB` may create a DLL with extension `.sl` on Unix. As the DLL file name is created using `tempfile()`, I think removing this corner case does not hurt, the file names should be unique anyways so that the object created should always have extension `.so`.

BTW, I was pretty much lost in trying to fix this until I discovered the awesome little helper function `cow.r`!

### CODE REVIEW

**Comment by eddelbuettel** [comment]
> so that the object created should always have extension .so.

That may not be correct. I think R itself as a function to determine the proper ending; code in `R CMD SHLIB` may reveal that.

The `tempfile` function can take a suggested ending [per my patch to R years ago ;-)] so maybe we can take that?

**Comment by eddelbuettel** [comment]
Oh, and now that I look at the code I see you use that via `.Platform$dynlib.ext`. Excellent.  This may be a minor 'smell' in use of `paste()`:

```r
libLFile <- file.path(tempdir(), paste0(f, .Platform$dynlib.ext))
```

where

```r
libLFile <- tempfile(pattern="link", fileext=.Platform$dynlib.ext) 
```

may be one better.  I will just make the change.

*Edit* Didn't work. Oh well.

File: R/cfunction.R [hunk 1]
```diff
@@ -275,29 +275,16 @@ compileCode <- function(f, code, language, verbose) {
...
-    ## UNIX-alike build
-    libCFile  <- paste(tempdir(), "/", f, ".EXT",               sep="")
-    libLFile  <- paste(tempdir(), "/", f, .Platform$dynlib.ext, sep="")
-    libLFile2 <- paste(tempdir(), "/", f, ".sl",                sep="")
-  }
   extension <- switch(language, "C++"=".cpp", C=".c", Fortran=".f", F95=".f95",
                                 ObjectiveC=".m", "ObjectiveC++"=".mm")
-  libCFile <- sub(".EXT$", extension, libCFile)
+  libCFile <- file.path(tempdir(), paste0(f, extension))
+  libLFile <- file.path(tempdir(), paste0(f, .Platform$dynlib.ext))
```
Comment (eddelbuettel) [inline R/cfunction.R:281]: That is a really nice cleanup. We can do one better which I'll add in a sec.

Comment (eddelbuettel) [reply R/cfunction.R:281]: Maybe not. Changing it I borked it, Guess I better leave it alone.

**Review by eddelbuettel — APPROVED** [review]

**Comment by eddelbuettel** [comment]
Looks good now. R-devel and r-release win-builder came back already with 'OK'.  Let's rest the thing for a few hours and see if anything comes.  Otherwise I will wrap it up as 0.3.17 tomorrow and ship.

**Comment by jranke** [comment]
OK, fingers crossed that MacOS will not gives us headaches...

**Comment by eddelbuettel** [comment]
For that we have `com.r` and `c4c.r` ;-)

**Comment by jranke** [comment]
Ah, `com.r` had escaped my attention. `c4c.r` only checked on two Linux flavors and Windows when I tried it. Nice work.

**Comment by eddelbuettel** [comment]
:)  `c4c` should be mac + win + linux, and if compiled code, also linux 'asan'.  RHub does the dispatching, not me.  You may have overlooked the mac there.  It's all still imperfect in the sense of build depends possibly missing etc

**Comment by eddelbuettel** [comment]
Grrrrr.  And just as I thought I had everything read, RHub (windows) gets me

```sh
  Running test_utilities.R..............    8 tests 1 fails Error in .local(x, ...) :
    Failed to copy DLL from C:\Users\USERwiEpaunUPL\AppData\Local\Temp\RtmpcxZesH/file18c06bae854.dll to C:\Users\USERwiEpaunUPL\AppData\Local\Temp\RtmpcxZesH/testname.dll
  Calls: <Anonymous> ... FUN -> eval -> eval -> moveDLL -> moveDLL -> .local
  In addition: Warning message:
  In file.create(to[okay]) :
    cannot create file 'C:\Users\USERwiEpaunUPL\AppData\Local\Temp\RtmpcxZesH/testname.dll', reason 'Permission denied'
  Execution halted
```

**Comment by eddelbuettel** [comment]
Adding `winslash="/"` to `normalizePath()` does nothing, this may be related to running DLLs.

```sh
 310#> Calls: ... FUN -> eval -> eval -> moveDLL -> moveDLL -> .local
 311#> Failed to copy DLL from C:\Users\USERQoAhFcvsYf\AppData\Local\Temp\RtmpmcjpbE/file278425fb5424.dll to C:/Users/USERQoAhFcvsYf/AppData/Local/Temp/RtmpmcjpbE/testname.dll
 312#> In addition: Warning message:
 313#> cannot create file 'C:/Users/USERQoAhFcvsYf/AppData/Local/Temp/RtmpmcjpbE/testname.dll', reason 'Permission denied'
 314#> Execution halted
```

**Comment by eddelbuettel** [comment]
So I propose to release with the following change skipping the "interesting" tests on Windows:

```diff
modified   inst/tinytest/test_utilities.R
@@ -1,5 +1,7 @@
 library(inline)
 
+onWindows <- .Platform$OS.type == "windows"
+
 code <- "
       int i;
       for (i = 0; i < *n; i++)
@@ -28,6 +30,8 @@ quadfn <- cfunction(signature(n = "integer", x = "numeric"), code,
 moveDLL(quadfn, name = "testname", directory = tempdir())
 expect_identical(quadfn(5, 1:5), res_known)
 
+if (onWindows) exit_file("Omitting remainder on Windows.")
+
 expect_error(
   moveDLL(quadfn, name = "testname", directory = tempdir()),
   "DLL .* in use")

```

That also means we actually don't really know if `moveDLL()` works on Windows.  That may be a problem for you in [mkin](https://cloud.r-project.org/web/packages/mkin/index.html).

But should of getting help from a Windows expert (hello @aadler -- could you possibly help us?) we're stuck here. 

[ @aadler In case you have a moment. This is package `inline` combining two of your favourites: Fortran (but also C and C++) and Windows.  On Windows moving/creating a file seems to fail about 5 out of 6 times. You can run the test file in question in isolation via  `Rscript -e 'tinytest::run_test_file("inst/tinytest/test_utilities.R")` ]
